contracts/EtomicSwapV2.sol

```diff
@@ -0,0 +1,476 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract EtomicSwapV2 {
+    using SafeERC20 for IERC20;
+
+    enum MakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerSpent,
+        MakerRefunded
+    }
+
+    struct MakerPayment {
+        bytes20 paymentHash;
+        uint32 paymentLockTime;
+        MakerPaymentState state;
+    }
+
+    event MakerPaymentSent(bytes32 id);
+    event MakerPaymentSpent(bytes32 id, bytes32 secret);
+    event MakerPaymentRefundedTimelock(bytes32 id);
+    event MakerPaymentRefundedSecret(bytes32 id);
+
+    mapping(bytes32 => MakerPayment) public makerPayments;
+
+    enum TakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerApproved,
+        MakerSpent,
+        TakerRefunded
+    }
+
+    struct TakerPayment {
+        bytes20 paymentHash;
+        uint32 immediateRefundTime;
+        uint32 paymentLockTime;
```

Comment:
Great note! I will do it shortly :slightly_smiling_face:

Reply:
```sol
 require( 
     block.timestamp >= 3 * takerPayments[id].paymentLockTime, 
     "Current timestamp didn't exceed payment refund lock time" 
 ); 
```
The above code is not accurate of course since paymentLockTime is a timestamp not duration :)